include/miopen/config.h.in

```diff
@@ -78,12 +78,55 @@
 
 // clang-format off
 #cmakedefine MIOPEN_DEFAULT_FIND_MODE @MIOPEN_DEFAULT_FIND_MODE@
+
 // clang-format on
+#ifndef MIOPEN_CMP_FLAGS_NEW
 #cmakedefine MIOPEN_AMDGCN_ASSEMBLER "@MIOPEN_AMDGCN_ASSEMBLER@"
 #cmakedefine HIP_OC_COMPILER "@HIP_OC_COMPILER@"
 #cmakedefine MIOPEN_HIP_COMPILER "@MIOPEN_HIP_COMPILER@"
 #cmakedefine MIOPEN_OFFLOADBUNDLER_BIN "@MIOPEN_OFFLOADBUNDLER_BIN@"
 #cmakedefine MIOPEN_CACHE_DIR "@MIOPEN_CACHE_DIR@"
+#else /*MIOPEN_CMP_FLAGS_NEW if defined*/
+/* Include rocm-core header for get ROCm Install Path Method */
+#include <rocm-core/rocm_getpath.h>
+
+/* Flags to help identify/differentiate Definitions for each MIOPEN Compiler Flags */
+typedef enum {
+        MIOPEN_AMDGCN_Flag=0,
+        MIOPEN_OC_COMPILER_Flag=1,
+        MIOPEN_HIP_COMPILER_Flag=2,
+        MIOPEN_OFFLOAD_BUNDLER_Flag=3,
+        MIOPEN_ROCM_INSTALL_PATH_Flag=4,
+        MIOPEN_CMP_MAX_Flag=5 /*Should be last Flag definition*/
+} MIOPEN_CMP_Flags_t;
+
+/* List to hold Relative Directory Path for each Compiler Flags from ROCM Install Path */
+extern const char* MIOPEN_CMP_FLAG_REL_PATHS[MIOPEN_CMP_MAX_Flag];
+
+/* List to hold the Compiler Flag Paths*/
+extern char MIOPEN_CMP_FLAGS[MIOPEN_CMP_MAX_Flag][PATH_MAX];
+
+/* Macro to initialize the MIOPEN Compiler Flags */
+#define MIOPEN_CMP_FLAGS_INIT() do{char *flagPath=NULL; int len = 0;  MIOPEN_CMP_Flags_t i;\
+        if(0==MIOPEN_CMP_FLAGS[MIOPEN_AMDGCN_Flag][0]){\
+          PathErrors_t ret = getROCmInstallPath(&flagPath, &len); \
+          if( PathSuccess == ret ){\
+            for(i=MIOPEN_AMDGCN_Flag;i<MIOPEN_CMP_MAX_Flag;i++){\
+                  sprintf((char *)*(MIOPEN_CMP_FLAGS+i),"%s%s", flagPath, MIOPEN_CMP_FLAG_REL_PATHS[i] );\
+            }\
+            free(flagPath);\
+          }\
+        }\
+       }while(0);
+
+
+#define MIOPEN_AMDGCN_ASSEMBLER MIOPEN_CMP_FLAGS[MIOPEN_AMDGCN_Flag]
```

Comment:
If you are going to get creative with macros, I would do it here.
Something like
#define get_cached_value(name) ( MIOPEN_CMP_FLAGS(name)?MIOPEN_CMP_FLAGS[name]:read_cache(name) )
#define MIOPEN_AMDGEN_ASSEMBLER get_cached(MIOPEN_AMDGEN_Flag)
#define MIOPEN_ROCM_INSTALL_PATH get_cache(MIOPEN_INSTALL_PATH_Flag)
and then implement read_cache to take in the enum and populate  the entry in MIOPEN_CMP_FLAGS and return it. This could be a normal function.

This approach means smaller code being copied into multiple .o files. Given something like MIOPEN_ROCM_INSTALL_PATH, it should compile into something like
load a value from a memory location
compare it to 0
jump if not equal
load a constant into a register
call read_cache

so 3 instructions in the normal case and gives you transparent caching. Note this is an alternative approach to having an "_INIT" style function.

Not super happy about the names of things. "Flags" to me suggests bool variables, rather than pathnames (strings)

Reply:
Thanks for the comments, Latest version updated with avoiding need for INIT invocation.